spopt/locate/p_median.py

```diff
@@ -533,3 +535,576 @@ def solve(self, solver: pulp.LpSolver, results: bool = True):
+                raise SpecificationError(
+                    "Problem is infeasible. The highest possible capacity "
+                    f"{highest_possible_capacity}, coming from the {p_facilities} "
+                    "sites with the highest capacity, is smaller than "
+                    f"the total demand {weights_sum}."
+                )
+            for j in col_indices:
+                model += (
+                    pulp.lpSum(
+                        weights[i] * cli_assgn_vars.get((i, j), 0) for i in r_cli
+                    )
+                    <= fac_vars[j] * facility_capacities[j]
+                )
+
+        # Create assignment constraints.
+        for i in r_cli:
+            model += (
+                pulp.lpSum(cli_assgn_vars.get((i, j), 0) for j in set(col_indices))
+                + placeholder_vars[i]
+                == 1
+            )
+        # Create the facility constraint.
+        FacilityModelBuilder.add_facility_constraint(k_nearest_p_median, p_facilities)
+
+        return k_nearest_p_median
+
+    @classmethod
+    def from_geodataframe(
+        cls,
+        k_list: np.array,
+        gdf_demand: GeoDataFrame,
+        gdf_fac: GeoDataFrame,
+        demand_col: str,
+        facility_col: str,
+        weights_cols: str,
+        p_facilities: int,
+        facility_capacity_col: str = None,
+        distance_metric: str = "euclidean",
+        name: str = "k-nearest-p-median",
+    ):
+        """
+        Set class variables for KNearestPMedian using input data.
+
+        Parameters
+        ----------
+        k_list : np.array
+            An array of integers representing the list of k values for each client.
+        gdf_demand : GeoDataFrame
+            A GeoDataFrame containing demand points with their associated attributes.
+        gdf_fac : GeoDataFrame
+            A GeoDataFrame containing facility points with their associated attributes.
+        demand_col : str
+            The column name in gdf_demand representing the coordinate of each client.
+        facility_col : str
+            The column name in gdf_fac representing the coordinate of each facility.
+        weights_cols : str
+            The column name in gdf_demand representing the weights for each client.
+        p_facilities: int
+           The number of facilities to be located.
+        facility_capacity_col : str, optional
+            The column name in gdf_fac representing the capacity of each facility,
+            by default None.
+        distance_metric : str, optional
+            The distance metric to be used in calculating distances between clients
+            and facilities, by default "euclidean".
+        name : str, optional
+            The name of the problem, by default "k-nearest-p-median".
+
+        Returns
+        -------
+        KNearestPMedian
+            The KNearestPMedian class itself with new class variables.
+
+        Warnings
```

Comment:
Are warnings thrown or are exceptions raised? If these are errors/exceptions this section title would probably be `Exceptions`